### ISSUE TITLE
Spawn entities with an `EntityCommand`-based abstraction

### ISSUE THREAD

**Issue by benfrankel** [TheBevyFlock/bevy_new_2d#223]
An example of how the proposed API looks in practice: https://github.com/benfrankel/bevy_jam_template/blob/main/src/screen/title.rs.

Benefits:
- You get an `EntityCommands` back immediately.
- You can access `SystemParam`s ergonomically by adding them as arguments.
- You can pass in extra data (e.g. `struct Projectile { lifetime, .. }`).
- Converting a "pure" entity spawning system to one that takes in extra data is relatively easy:
    - Define your data type, like `struct Projectile`.
    - Implement `EntityCommand` for `Projectile`, and delegate to the spawning system with `world.run_system_once_with((id, self), projectile)`.
    - Change `In(id): In<Entity>` to `In((id, this)): In<(Entity, Projectile)>`.

Drawbacks:
- Requires extension traits.
- Unfamiliar approach (using observers to spawn things is unfamiliar as well).
- `spawn_fn` and `spawn_with` unfortunately couldn't be merged into one method.
- A few lines of boilerplate to take in data. This could use a VS Code snippet to make it easier.

**Comment by janhohenheim** [comment]
I'm in favor.
Note that we went through many designs and this seems the cleanest by far imo. Bevy does not have the required answer to "How do I spawn an object composed of lots of stuff?" yet while still returning an `EntityCommands`, so we *need* to introduce an abstraction here.

**Comment by alice-i-cecile** [comment]
I'm also in favour of this.

**Comment by janhohenheim** [comment]
Blocked on upstreaming https://github.com/bevyengine/bevy/issues/14278 and maybe https://github.com/bevyengine/bevy/issues/14231

**Comment by benfrankel** [comment]
Closing this since it's definitely not going to happen in the form originally proposed. We've moved on to Bevy 0.16 spawning API, which still has issues but should hopefully improve further in the coming releases.
